core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -350,7 +328,261 @@ public void run() {
     }
   }
 
-  static class NoopClientStream implements ClientStream {
+  /**
+   * A stream that queues requests before the transport is available, and delegates to a real stream
+   * implementation when the transport is available.
+   *
+   * <p>{@code ClientStream} itself doesn't require thread-safety. However, the state of {@code
+   * DelayedStream} may be internally altered by different threads, thus internal synchronization is
+   * necessary.
+   */
+  private class DelayedStream implements ClientStream {
+    final Metadata headers;
+    final ClientStreamListener listener;
+    final ListenableFuture<ClientTransport> transportFuture;
+
+    // volatile to be readable without synchronization in the fast path
+    @GuardedBy("this")
+    volatile ClientStream realStream;
+
+    @GuardedBy("this")
+    Compressor compressor;
+    // Can be either a Decompressor or a String
+    @GuardedBy("this")
+    Object decompressor;
+    @GuardedBy("this")
+    DecompressorRegistry decompressionRegistry;
+    @GuardedBy("this")
+    final LinkedList<InputStream> pendingMessages = new LinkedList<InputStream>();
+    @GuardedBy("this")
+    boolean pendingHalfClose;
+    @GuardedBy("this")
+    int pendingFlowControlRequests;
+    @GuardedBy("this")
+    boolean pendingFlush;
+
+
+    /**
+     * Get a transport and try to create a stream on it.
+     */
+    private final Runnable streamCreationTask = new Runnable() {
+      @Override
+      public void run() {
+        // Fill out timeout on the headers
+        headers.removeAll(TIMEOUT_KEY);
+        // Convert the deadline to timeout. Timeout is more favorable than deadline on the wire
+        // because timeout tolerates the clock difference between machines.
+        Long timeoutMicros = getRemainingTimeoutMicros();
+        if (timeoutMicros != null) {
+          if (timeoutMicros <= 0) {
+            maybeClosePrematurely(Status.DEADLINE_EXCEEDED);
+            return;
+          }
+          headers.put(TIMEOUT_KEY, timeoutMicros);
+        }
+
+        if (transportFuture.isDone()) {
+          ClientTransport transport;
+          try {
+            transport = transportFuture.get();
+          } catch (Exception e) {
+            maybeClosePrematurely(Status.fromThrowable(e));
+            return;
+          }
+          createStream(transport);
+        } else {
+          transportFuture.addListener(streamCreationTask, callExecutor);
```

Comment:
I expect we will want this handled outside of DelayedStream since if the transportFuture is done initially then there is no need to use a DelayedStream.